### DISCUSSION TITLE
Jemalloc on Gramine

### DISCUSSION THREAD

**Discussion by ila** [gramineproject/gramine#1825] [Performance discussions]
Hi, we finally managed to obtain good OLAP performance on SGX2 with Gramine. I am happy to announce that we get maximum 2x overhead with medium-sized, memory-intensive workloads (TPC-H): this is thanks to careful configuration of Gramine and our application, NUMA pinning and hardware-optimized instructions. 

However, there is one last thing we cannot explain fully: the performance of jemalloc (https://jemalloc.net/) in Gramine is terrible (16x overhead), we had to remove it. We are working under the condition that `data size < enclave size < EPC size`, to minimize page swapping. So, we can (probably?) exclude spilling.

What could the cause be? As far as I can tell, jemalloc is specifically designed to avoid fragmentation (one of the enemies of SGX too). Yet, we see the opposite behaviour. Is this because Gramine does not implement its own jemalloc? Could it be that malloc is more optimized in Gramine? Or do you have any other idea? Logs aren't much useful here. Any input would be appreciated!

**Comment by mkow** [comment]
Could you try running it under a profiler and compare the traces with jemalloc and without? See https://gramine.readthedocs.io/en/latest/performance.html#profiling-with-perf.

**Reply by ila** [reply]
Sorry for delay, only going through backlog now.

With `jemalloc`:

```
name    run     timing
benchmark/tpch/sf30_cu/q18.benchmark    1       15.700586
benchmark/tpch/sf30_cu/q18.benchmark    2       20.433170
benchmark/tpch/sf30_cu/q18.benchmark    3       20.081889
benchmark/tpch/sf30_cu/q18.benchmark    4       25.151581
benchmark/tpch/sf30_cu/q18.benchmark    5       16.148420

 Performance counter stats for 'gramine-sgx benchmark_runner benchmark/tpch/sf30_cu/q18.benchmark':

       49587089531      branches                         #   13.132 M/sec                       (90.00%)
         168481665      branch-misses                    #    0.34% of all branches             (90.00%)
      381953248875      bus-cycles                       #  101.155 M/sec                       (89.99%)
         715957736      cache-misses                     #   69.37% of all cache refs           (90.00%)
        1032062736      cache-references                 #  273.327 K/sec                       (90.00%)
      431774336511      cycles                           #    0.114 GHz                         (80.00%)
      230891209836      instructions                     #    0.53  insn per cycle              (90.00%)
        3775925.58 msec cpu-clock                        #   11.981 CPUs utilized             
               542      cpu-migrations                   #    0.144 /sec                      
          16793721      page-faults                      #    4.448 K/sec                     
        2942326580      L1-dcache-load-misses                                                   (90.01%)
         161306184      LLC-load-misses                                                         (90.01%)
          80164117      LLC-store-misses                 #   21.230 K/sec                       (90.00%)

     315.152281765 seconds time elapsed

    3628.313429000 seconds user
     148.833983000 seconds sys

```

Without `jemalloc`:
```
name    run     timing
benchmark/tpch/sf30_cu/q18.benchmark    1       2.293231
benchmark/tpch/sf30_cu/q18.benchmark    2       2.506912
benchmark/tpch/sf30_cu/q18.benchmark    3       2.261930
benchmark/tpch/sf30_cu/q18.benchmark    4       2.650745
benchmark/tpch/sf30_cu/q18.benchmark    5       2.092871

 Performance counter stats for 'gramine-sgx benchmark_runner benchmark/tpch/sf30_cu/q18.benchmark':

       48552969951      branches                         #   59.083 M/sec                       (90.01%)
         137121102      branch-misses                    #    0.28% of all branches             (90.02%)
      319907455636      bus-cycles                       #  389.290 M/sec                       (90.00%)
         461773353      cache-misses                     #   69.28% of all cache refs           (90.01%)
         666531531      cache-references                 #  811.091 K/sec                       (89.99%)
      412177976912      cycles                           #    0.502 GHz                         (80.00%)
      224949864702      instructions                     #    0.55  insn per cycle              (90.01%)
         821771.15 msec cpu-clock                        #    4.158 CPUs utilized             
               728      cpu-migrations                   #    0.886 /sec                      
          16794648      page-faults                      #   20.437 K/sec                     
        3110913772      L1-dcache-load-misses                                                   (90.01%)
         142318786      LLC-load-misses                                                         (90.01%)
          31913590      LLC-store-misses                 #   38.835 K/sec                       (90.02%)

     197.651681003 seconds time elapsed

     690.095148000 seconds user
     131.299220000 seconds sys
```

It looks like the stats are very different for the very same workload. The cache misses are expected, since this workload involves a lot of hash tables - however I can't explain the other statistics. It just looks like the `jemalloc` version is more compute-intensive, so it could lead to more cache misses?

**Reply by dimakuv** [reply]
@ila It would be interesting to see the statistics for the original application itself, so that we have 4 points of comparison.

Looking at these perf stats, the first thing that jumps out is the CPU caches utilization. The no-jemalloc version of Gramine-SGX has significantly less cache usage. However, there is a lot of info missing in these stats, I don't think we can conclude anything with 100% clarity.

I would recommend doing more perf analysis:
- Run `perf stat` with TLB counters too (in all 4 versions)
- Run the original application under `strace -f -c`, so that we know the syscall behavior of the app (in 2 versions -- with and without jemalloc)
- Run with `sgx.enable_stats = true` (in 2 Gramine-SGX versions -- with and without jemalloc) and show us the "Total stats" on EENTER, EEXIT, AEX
- Advanced level: run `perf record`-style Gramine profiler (see the link provided by mkow) and show us `perf report` results (again, ideally in all 4 versions)

Anyhow, `jemalloc` is typically a boost for applications and for Gramine-SGX. We've seen only good results with it so far. I'm curious why it behaves badly in your application.

**Reply by ila** [reply]
Thanks for the reply. Here are some preliminary benchmarks, will continue tomorrow (need to rebuild everything in debug mode). It looks like `jemalloc` is slightly slower on our application too, although there is some noise caused by measuring the performance counters.

You mention that `jemalloc` boosts performance, do you have some examples? Some applications I could test?

Results:

### strace (jemalloc)
```
benchmark/tpch/sf30_cu/q18.benchmark   1    strace: Process 1324908 attached 
8.459757 
benchmark/tpch/sf30_cu/q18.benchmark   2    strace: Process 1324909 attached 
7.762700 
benchmark/tpch/sf30_cu/q18.benchmark   3    strace: Process 1324921 attached 
8.060001 
benchmark/tpch/sf30_cu/q18.benchmark   4    strace: Process 1324922 attached 
7.814841 
benchmark/tpch/sf30_cu/q18.benchmark   5    strace: Process 1324934 attached 
7.768518 
% time   seconds  usecs/call   calls   errors syscall 

------ ----------- ----------- --------- --------- ------------------

 45.92 1737.006201     1629  1066176      pread64 
 36.53 1381.962891    30947   44655    6723 futex 
  7.01  265.178195     3929   67476      munmap 
  5.12  193.784056     2748   70517      mmap 
  3.41  129.062438    27827    4638      clock_nanosleep 
  1.90  72.057930     3797   18976      mprotect 
  0.03   1.061621     231    4585     3 newfstatat 
  0.02   0.898081     626    1434      close 
  0.02   0.876966     608    1441     7 openat 
  0.02   0.600632     1201    500      madvise 
  0.01   0.241840     767    315      brk 
  0.00   0.115467     101    1137      rt_sigprocmask 
  0.00   0.043085     151    284      clone3 
  0.00   0.032720     114    285      set_robust_list 
  0.00   0.024766      86    285      rseq 
  0.00   0.023103      7    3061     3 access 
  0.00   0.004232      9    466      getdents64 
  0.00   0.000150      13     11      write 
  0.00   0.000063      4     13      read 
  0.00   0.000035      35     1     1 readlink 
  0.00   0.000019      6     3      sched_getaffinity 
  0.00   0.000010      3     3      lseek 
  0.00   0.000009      9     1      chdir 
  0.00   0.000009      9     1      prlimit64 
  0.00   0.000008      8     1      getrandom 
  0.00   0.000007      7     1      rt_sigaction 
  0.00   0.000007      7     1      sysinfo 
  0.00   0.000000      0     1      execve 
  0.00   0.000000      0     2     1 arch_prctl 
  0.00   0.000000      0     1      set_tid_address 

------ ----------- ----------- --------- --------- ------------------

100.00 3782.974541     2941  1286271    6738 total
```

### strace (malloc)
```
benchmark/tpch/sf30_cu/q18.benchmark   1    strace: Process 1311061 attached 
6.519128 
benchmark/tpch/sf30_cu/q18.benchmark   2    strace: Process 1311073 attached 
6.744344 
benchmark/tpch/sf30_cu/q18.benchmark   3    strace: Process 1311077 attached 
6.570212 
benchmark/tpch/sf30_cu/q18.benchmark   4    strace: Process 1311078 attached 
6.756836 
benchmark/tpch/sf30_cu/q18.benchmark   5    strace: Process 1311091 attached 
6.867340 
% time   seconds  usecs/call   calls   errors syscall 

------ ----------- ----------- --------- --------- ------------------

 69.75 1861.558788     1746  1066176      pread64 
 24.58  655.999525    18381   35688    7356 futex 
  4.13  110.255320    28445    3876      clock_nanosleep 
  1.21  32.402032     1021   31735      mprotect 
  0.11   3.000447     6123    490      munmap 
  0.04   1.027050     700    1466     7 openat 
  0.04   1.005068     218    4610     3 newfstatat 
  0.04   0.980228     2883    340      madvise 
  0.04   0.974532     667    1459      close 
  0.03   0.820949     1600    513      mmap 
  0.02   0.596721     689    865      brk 
  0.00   0.122857     108    1137      rt_sigprocmask 
  0.00   0.034169     120    284      clone3 
  0.00   0.028966     101    285      set_robust_list 
  0.00   0.024896      87    285      rseq 
  0.00   0.020163      6    3061     3 access 
  0.00   0.003697      7    466      getdents64 
  0.00   0.000336     168     2      sched_getaffinity 
  0.00   0.000083      6     13      read 
  0.00   0.000044      4     11      write 
  0.00   0.000021      7     3      lseek 
  0.00   0.000007      7     1      chdir 
  0.00   0.000007      7     1      sysinfo 
  0.00   0.000003      3     1      rt_sigaction 
  0.00   0.000000      0     1      execve 
  0.00   0.000000      0     2     1 arch_prctl 
  0.00   0.000000      0     1      set_tid_address 
  0.00   0.000000      0     1      prlimit64 
  0.00   0.000000      0     1      getrandom 

------ ----------- ----------- --------- --------- ------------------

100.00 2668.855909     2315  1152774    7370 total
```

### perf stat (jemalloc)
```
Performance counter stats for '../duckdb_openssl/build/release/benchmark/benchmark_runner benchmark
/tpch/sf30_cu/q18.benchmark': 

    219490123744    branches             #  453.605 M/sec            (35.

84%) 
     1588989684    branch-misses           #   0.72% of all branches       (42.
98%) 
   1003979790550    bus-cycles            #   2.075 G/sec            (50.
11%) 
     4843185434    cache-misses           #  61.58% of all cache refs      (57.
22%) 
     7865210958    cache-references         #  16.254 M/sec            (57.
21%) 
    645511069499    cycles              #   1.334 GHz             (57.
21%) 
   1176580630824    instructions           #   1.82  insn per cycle        (64.
35%) 
     483879.28 msec cpu-clock             #  23.143 CPUs utilized        
        740    cpu-migrations          #   1.529 /sec            
      7365294    page-faults            #  15.221 K/sec            
    18260981813    L1-dcache-load-misses                          (64.
35%) 
     2498685687    LLC-load-misses                             (64.
29%) 
     568280495    LLC-store-misses         #   1.174 M/sec            (64.
25%) 
    293945499626    dTLB-loads            #  607.477 M/sec            (64.
28%) 
     2518159204    dTLB-load-misses         #   0.86% of all dTLB cache accesses  (28.
54%) 
    112198711823    dTLB-stores            #  231.873 M/sec            (28.
63%) 
     284470962    dTLB-store-misses         #  587.897 K/sec            (28.
65%) 

    20.907867662 seconds time elapsed 

   424.837159000 seconds user 
    59.727474000 seconds sys
```

### perf stat (jemalloc, SGX)
```
----- Total SGX stats for process 1323405 ----- 
  \# of EENTERs:     1300102 
  \# of EEXITs:     1299417 
  \# of AEXs:      17403543 
  \# of sync signals:  111 
  \# of async signals:  0 

 Performance counter stats for 'gramine-sgx benchmark_runner benchmark/tpch/sf30_cu/q18.benchmark': 

    413241598470    branches             #  160.244 M/sec            (35.

77%) 
     2134302869    branch-misses           #   0.52% of all branches       (42.
91%) 
   5388090077274    bus-cycles            #   2.089 G/sec            (50.
05%) 
     8897326642    cache-misses           #  66.81% of all cache refs      (57.
18%) 
    13317999151    cache-references         #   5.164 M/sec            (57.
17%) 
   5060780975903    cycles              #   1.962 GHz             (57.
18%) 
   1765505806691    instructions           #   0.35  insn per cycle        (64.
32%) 
     2578831.71 msec cpu-clock             #  10.907 CPUs utilized        
        694    cpu-migrations          #   0.269 /sec            
      16794247    page-faults            #   6.512 K/sec            
    26690441068    L1-dcache-load-misses                          (64.
30%) 
     4492183303    LLC-load-misses                             (64.
30%) 
     1265160830    LLC-store-misses         #  490.595 K/sec            (64.
30%) 
    419689310124    dTLB-loads            #  162.744 M/sec            (64.
30%) 
     6061607978    dTLB-load-misses         #   1.44% of all dTLB cache accesses  (28.
58%) 
    133108057363    dTLB-stores            #  51.616 M/sec            (28.
61%) 
     704165633    dTLB-store-misses         #  273.056 K/sec            (28.
61%) 

   236.431820308 seconds time elapsed 

   2464.866123000 seconds user 
   114.642238000 seconds sys
```

### perf stat (malloc)
```
 Performance counter stats for '../duckdb_openssl/build/release/benchmark/benchmark_runner benchmark
/tpch/sf30_cu/q18.benchmark': 

    217662914953    branches             #  564.990 M/sec            (35.

78%) 
     1546828751    branch-misses           #   0.71% of all branches       (42.
91%) 
    801120171628    bus-cycles            #   2.079 G/sec            (50.
07%) 
     4913267051    cache-misses           #  67.13% of all cache refs      (57.
19%) 
     7318915144    cache-references         #  18.998 M/sec            (57.
19%) 
    604257896132    cycles              #   1.568 GHz             (57.
22%) 
   1164492060601    instructions           #   1.93  insn per cycle        (64.
39%) 
     385251.09 msec cpu-clock             #  25.468 CPUs utilized        
        514    cpu-migrations          #   1.334 /sec            
      6669153    page-faults            #  17.311 K/sec            
    18530047580    L1-dcache-load-misses                          (64.
37%) 
     2725451675    LLC-load-misses                             (64.
38%) 
     582243721    LLC-store-misses         #   1.511 M/sec            (64.
37%) 
    290482141230    dTLB-loads            #  754.007 M/sec            (64.
36%) 
     2347972258    dTLB-load-misses         #   0.81% of all dTLB cache accesses  (28.
56%) 
    110427953112    dTLB-stores            #  286.639 M/sec            (28.
61%) 
     220769612    dTLB-store-misses         #  573.054 K/sec            (28.
62%) 

    15.126682627 seconds time elapsed 

   339.097192000 seconds user 
    46.398703000 seconds sys
```

### perf stat (malloc, SGX)
```
----- Total SGX stats for process 1309159 ----- 
  \# of EENTERs:     1137525 
  \# of EEXITs:     1136840 
  \# of AEXs:      16986518 
  \# of sync signals:  111 
  \# of async signals:  0 

 Performance counter stats for 'gramine-sgx benchmark_runner benchmark/tpch/sf30_cu/q18.benchmark': 

    261511430041    branches             #  283.765 M/sec            (35.84%) 
     1761982397    branch-misses           #   0.67% of all branches       (42.98%) 

   1922626516146    bus-cycles            #   2.086 G/sec            (50.11%) 
     8684797765    cache-misses           #  73.07% of all cache refs      (57.23%) 
    11885930493    cache-references         #  12.897 M/sec            (57.20%) 
   1893049531576    cycles              #   2.054 GHz             (57.21%) 
   1371971755063    instructions           #   0.72  insn per cycle        (64.36%) 
     921578.90 msec cpu-clock             #   4.953 CPUs utilized        
        501    cpu-migrations          #   0.544 /sec            
      16803973    page-faults            #  18.234 K/sec            
    22382443442    L1-dcache-load-misses                          (64.35%) 
     3473772078    LLC-load-misses                             (64.35%) 
     2005650849    LLC-store-misses         #   2.176 M/sec            (64.35%) 
    339521583825    dTLB-loads            #  368.413 M/sec            (64.34%) 
     5895241993    dTLB-load-misses         #   1.74% of all dTLB cache accesses  (28.59%) 
    131250847689    dTLB-stores            #  142.420 M/sec            (28.68%) 
     720711014    dTLB-store-misses         #  782.039 K/sec            (28.68%) 

   186.077661491 seconds time elapsed 

   794.621350000 seconds user 
   126.626145000 seconds sys
```

**Reply by dimakuv** [reply]
> You mention that jemalloc boosts performance, do you have some examples? Some applications I could test?

That was just a general feeling, I don't have any particular examples.

Looking at our recommendations, I see that `mimalloc` and `tcmalloc` are the best-performing memory allocators under Gramine-SGX. So you may want to try them: https://github.com/gramineproject/examples/tree/master/openvino#memory-allocator-libraries

---

```
### strace (jemalloc)

% time   seconds  usecs/call   calls   syscall 
------ ----------- ----------- --------- ------------------
  ...
  7.01  265.178195     3929   67476      munmap 
  5.12  193.784056     2748   70517      mmap 

### strace (malloc)

% time   seconds  usecs/call   calls   syscall 
------ ----------- ----------- --------- ------------------
  ...
  0.11   3.000447     6123    490      munmap 
  0.03   0.820949     1600    513      mmap 
```

The above are from your experiments on native execution. I think this snippet with usage of `mmap()` and `munmap()` explains it all -- `jemalloc` is very memory-allocations-intensive, whereas default `malloc` has only a few memory allocations.

It feels like your application is heavily optimized for `malloc` rather than `jemalloc`. **What is the perf difference on native application?** You only mentioned that Gramine-SGX performs significantly worse with `jemalloc` than with `malloc`, but what about your original application? Looking at these straces, the original app itself must perform worse with `jemalloc` than with `malloc`.

**Reply by ila** [reply]
Yes, we are still looking into it - I noticed a 10% performance drop when using `jemalloc` rather than `malloc`. I think our application could be optimized for `jemalloc` with ARM processors, and running it on Intel CPUs causes some weird behaviour (however, I am not the one who developed this, so I want to be extra sure before I bring this up). 

Or possibly, the slight performance difference is a tradeoff for less memory consumption in memory-intensive workloads. 

However, I still cannot explain this huge difference in SGX -- 10% overhead is tolerable, but 2s vs 20s in SGX is not. When the memory is allocated (ie, `mmap` is called), this should be inside the enclave, right? Could this overhead be the cost of zeroing/encrypting the pages? 

Currently trying to analyze `perf record` output, will keep you posted.

**Reply by dimakuv** [reply]
@ila Is Gramine-SGX running with EDMM enabled in your case? (`sgx.edmm_enable = true`)

EDMM is expensive due to how the in-Linux implementation currently works. So seeing 10x overhead is not that surprising when using EDMM.

And even without EDMM, each `mmap()` requires to zero out the memory region, which is also a rather slow operation.

Finally, we know of an inefficiently in memory allocations in Gramine. I submitted a patch to fix this: https://github.com/gramineproject/gramine/pull/1795. You can try this patch on your application, and see what difference it makes.

TLDR: In any case, memory allocations in SGX are expensive.

**Reply by ila** [reply]
No, we are not using EDMM. I understand that `mmap` zeroes memory, but as far as I know, this happens for security reasons -- for instance if process A uses memory, then frees it, then process B needs memory, the pages are zeroed to avoid process B accidentally reading whatever process A was doing.

Inside an enclave, allocating memory also inside an enclave, is this concern still valid? If we can trust our application and the processes accessing it, is still there a need to zero the pages? 

Thanks to the pointer to this patch! I will test it now.

**Reply by dimakuv** [reply]
> No, we are not using EDMM. I understand that `mmap` zeroes memory, but as far as I know, this happens for security reasons -- for instance if process A uses memory, then frees it, then process B needs memory, the pages are zeroed to avoid process B accidentally reading whatever process A was doing.

Not only for security reasons, but also to provide some sane default values in the freshly-allocated memory region. If the same process A first allocates `[0, 4096)` and writes all ones to it, then frees it, then allocates this region again, then it must have all zeros (and not the previously written all-ones).

> Inside an enclave, allocating memory also inside an enclave, is this concern still valid? If we can trust our application and the processes accessing it, is still there a need to zero the pages?

No, we can't skip zeroing out. See my explanation above -- this is a requirement that applications rely on.

However, I must note that normal OS kernels like Linux and Windows play special tricks to **not do** literal zero-out of pages. Instead, Linux puts a special all-zeros page into the page tables upon `mmap()` requests. Which is a much faster operation than `memset(0)`. I'm sure Windows has something similar.

SGX cannot do such a trick because SGX doesn't have access to page tables. So we must fall back to a more straight-forward and less efficient `memset(0)`.

**Reply by ila** [reply]
That's very informative, thanks a lot for your explanation! Unfortunately your patch doesn't really work (using `jemalloc`):

```
gramine-sgx --version
Gramine 1.6.1post~UNRELEASED (c208c3cd58ab34d36fd7c2d9e2b73e704153d5a7)

gramine-sgx benchmark_runner benchmark/tpch/sf30_cu/q18.benchmark
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
name    run     timing
benchmark/tpch/sf30_cu/q18.benchmark    1       7.308045
benchmark/tpch/sf30_cu/q18.benchmark    2       TIMEOUT
```

Without `jemalloc`, performance seems to also slower, compared to the very first  `perf` output I showed you:

(Edit: just tested with more benchmarks. It is indeed slower,  around 50-60%) 
```
perf stat -e branches,branch-misses,bus-cycles,cache-misses,cache-references,cycles,instructions,cpu-clock,cpu-migrations,page-faults,L1-dcache-load-misses,LLC-load-misses,LLC-store-misses,dTLB-loads,dTLB-load-misses,dTLB-stores,dTLB-store-misses gramine-sgx benchmark_runner "benchmark/tpch/sf30_cu/q18.benchmark"
-----------------------------------------------------------------------------------------------------------------------

name    run     timing
benchmark/tpch/sf30_cu/q18.benchmark    1       3.240562
benchmark/tpch/sf30_cu/q18.benchmark    2       3.295672
benchmark/tpch/sf30_cu/q18.benchmark    3       4.101344
benchmark/tpch/sf30_cu/q18.benchmark    4       3.080541
benchmark/tpch/sf30_cu/q18.benchmark    5       2.772859

 Performance counter stats for 'gramine-sgx benchmark_runner benchmark/tpch/sf30_cu/q18.benchmark':

       47699917097      branches                         #   58.602 M/sec                       (35.85%)
         135659034      branch-misses                    #    0.28% of all branches             (42.98%)
      272904831992      bus-cycles                       #  335.278 M/sec                       (50.11%)
         506351404      cache-misses                     #   71.32% of all cache refs           (57.23%)
         709986440      cache-references                 #  872.256 K/sec                       (57.21%)
      415787434382      cycles                           #    0.511 GHz                         (57.20%)
      222799482224      instructions                     #    0.54  insn per cycle              (64.35%)
         813965.65 msec cpu-clock                        #    4.772 CPUs utilized             
               800      cpu-migrations                   #    0.983 /sec                      
          16799594      page-faults                      #   20.639 K/sec                     
        2835159342      L1-dcache-load-misses                                                   (64.35%)
         152516370      LLC-load-misses                                                         (64.35%)
          60101682      LLC-store-misses                 #   73.838 K/sec                       (64.33%)
       52770855375      dTLB-loads                       #   64.832 M/sec                       (64.33%)
        2089194709      dTLB-load-misses                 #    3.96% of all dTLB cache accesses  (28.60%)
       24254876796      dTLB-stores                      #   29.798 M/sec                       (28.66%)
         296377771      dTLB-store-misses                #  364.116 K/sec                       (28.67%)

     170.566229709 seconds time elapsed

     703.499911000 seconds user
     110.423155000 seconds sys
```
